packages/babel-plugin-proposal-decorators/src/transformer-2021-12.ts

```diff
@@ -0,0 +1,954 @@
+    ]);
+
+    if (firstFieldPath) {
+      const value = firstFieldPath.get("value");
+      const body: t.Expression[] = [protoInitCall];
+
+      if (value.node) {
+        body.push(value.node);
+      }
+
+      value.replaceWith(t.sequenceExpression(body));
+    } else if (constructorPath) {
+      if (path.node.superClass) {
+        const superPath = constructorPath
+          .get("body.body")
+          .find(
+            path =>
+              path.node.type === "ExpressionStatement" &&
+              path.node.expression.type === "CallExpression" &&
+              path.node.expression.callee.type === "Super",
+          );
+
+        if (!superPath) {
+          throw new Error(
+            `No super found inside class constructor for ${
+              path.node.id ? path.node.id.name : "an anonymous class"
+            }, but the class extends from another class`,
+          );
+        }
+
+        superPath.insertAfter(t.expressionStatement(protoInitCall));
+      } else {
+        constructorPath.node.body.body.unshift(
+          t.expressionStatement(protoInitCall),
+        );
+      }
+    } else {
+      const body: t.Statement[] = [t.expressionStatement(protoInitCall)];
+
+      if (path.node.superClass) {
+        body.unshift(
+          t.expressionStatement(
+            t.callExpression(t.super(), [
+              t.spreadElement(t.identifier("args")),
+            ]),
+          ),
+        );
+      }
+
+      path.node.body.body.unshift(
+        t.classMethod(
+          "constructor",
+          t.identifier("constructor"),
+          [t.restElement(t.identifier("args"))],
+          t.blockStatement(body),
+        ),
+      );
+    }
+  }
+
+  if (requiresStaticInit) {
+    staticInitLocal = generateLocalVarId(path, "initStatic");
+    locals.push(staticInitLocal);
+  }
+
+  if (classDecorators) {
+    classInitLocal = generateLocalVarId(path, "initClass");
+
+    const [localId, classPath] = replaceClassWithVar(path);
+    path = classPath;
+
+    locals.push(localId, classInitLocal);
+    path.node.decorators = null;
+  }
+
+  const staticBlock = t.staticBlock(
+    [
+      t.expressionStatement(
+        t.assignmentExpression(
+          "=",
+          t.arrayPattern(locals),
+          t.callExpression(state.addHelper("applyDecs"), [
+            t.thisExpression(),
+            elementDecorations,
+            classDecorations,
+          ]),
+        ),
+      ),
+      requiresStaticInit &&
+        t.expressionStatement(
+          t.callExpression(t.cloneNode(staticInitLocal), [t.thisExpression()]),
+        ),
+    ].filter(v => v),
+  );
+
+  path.node.body.body.unshift(staticBlock as unknown as ClassElement);
+
+  if (classInitLocal) {
+    path.node.body.body.push(
+      t.staticBlock([
+        t.expressionStatement(
+          t.callExpression(t.cloneNode(classInitLocal), []),
+        ),
+      ]),
+    );
+  }
+
+  path.traverse({
+    AssignmentExpression(path) {
```

Comment:
Edge cases: A member expression can also appears in a UpdateExpression or a destructuring assignment target:

```js
this.#x++;
[...this.#x] = [];
for (this.#x of []);
```

Reply:
hmm, in those cases I don't believe that the `readOnlyError` helper will really work, it'll be a syntax error. Should we just throw a compiler error if we find any assignments to the method? This may also be a capability we add in the future (see: https://github.com/tc39/proposal-decorators/issues/438) so we could throw an error for now, and in the future if we end up supporting setting to decorated methods just remove that error.

Reply:
We already handle that case poorly (https://babeljs.io/repl#?browsers=chrome%2080&build=&builtIns=false&corejs=3.6&spec=false&loose=false&code_lz=MYGwhgzhAECC0G8BQ1oGIAeAKAlIgvitALYCmALgBYD2AJrokagNoB07VAlhK5gLrQAvNGZ8A3EUKEkAO1IB3OLlZkqdXEA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=env&prettier=false&targets=&version=7.16.6&externalPlugins=&assumptions=%7B%7D is wrong), so a compile-time error is fine for now.